### PR TITLE
✨ Add support negative polarity conditions

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR is an iteration on the condition proposal & corresponding utils allowing the implementation of conditions with negative polarity, which is a step toward latest K8s guidelines about conditions

**Which issue(s) this PR fixes**:
Part of https://github.com/kubernetes-sigs/cluster-api/issues/7635

/area api

cc @JoelSpeed @vincepri @theobarberbany

### CODE REVIEW

File: util/conditions/getter_test.go
Comment (JoelSpeed) [inline util/conditions/getter_test.go]: Guessing this one isn't needed?

File: util/conditions/setter.go
Comment (enxebre) [inline util/conditions/setter.go:111]: Why can't this have a reason and message?

File: util/conditions/setter.go
Comment (enxebre) [inline util/conditions/setter.go:100]: why do NegativeTrueCondition and PositiveFalseCondition include the polarity in the func name?
Since all they are doing is set values in a clusterv1.Condition type wouldn't they be called just TrueCondition / falseCondition and let higher invoker decide on polarity?

File: util/conditions/setter.go
Comment (fabriziopandini) [reply util/conditions/setter.go:111]: Back in time we introduced all those helpers to enforce a sort of consistency, and what we have to express when status is good (TrueCondition) does not allow to express reason and message & severity.
This func mirror the same for a condition with negative polarity(when status is False but with a positive meaning).

In other words, I tried to keep things consistent with what we have.

Comment (JoelSpeed) [reply util/conditions/setter.go:111]: Sometimes it can be awkward to require a reason, but there are cases where I think it can make sense.

I guess something like Available as a condition, True status, reason would be AsExpected maybe? But does it also make sense without?

But yeah, in general, I wonder if we should just have the reason and message allowed in all cases, rather than being prescriptive?

Are there kube conventions that say you should or shouldn't set reason/message in certain cases?

File: util/conditions/setter.go
Comment (fabriziopandini) [reply util/conditions/setter.go:100]: See https://github.com/kubernetes-sigs/cluster-api/pull/10550#discussion_r1588979694, but the intent is to keep things consistent with what we have.

Just a note about the approach you are suggesting about let higher invoker decide on polarity: this is not in discussion.

The difference is that:
- If we change the existing TrueCondition & FalseCondition, this will impact everyone and we end up to defer to users the task to enforce the same consistency that today we have with this helpers
- With the proposed change instead, the users specific helpers for the NegativePolarity, which are consistent with the existing one for positive polarity

Last, but not least, there is already a Set helper with allow the users to decide which field to set (but in this case, it also imply the task to enforce consistency across the codebase/all the providers)

File: util/conditions/setter.go
Comment (fabriziopandini) [reply util/conditions/setter.go:111]: If a user wants flexibility, he can already do this today using the Set helper.
But as far as I know, in the last 4 years almost everyone used "prescriptive" helpers.

WRT to API conventions, before digging into "reason", let's keep in mind that the scope of this PR is to make one step toward API compliance, not the entire path 😉 

Based on my understanding, reason is another area where the thinking about convention evolved over time. The current field description in corev1 types sort of captures it:

> // reason contains a programmatic identifier indicating the reason for the condition's last transition.
> // Producers of specific condition types may define expected values and meanings for this field,
> // and whether the values are considered a guaranteed API.
> // The value should be a CamelCase string.
> // This field may not be empty.

My reading of this is that community is moving towards having reason everywhere, but they recognize producers of condition might have adopted different approaches.

CAPI as of today has two main differences from the API conventions WRT to how we handle reasons:
- From a semantic point of view, we use reason to "explain" the current state, not the last transition (in some case the diff is subtle, but in controller we just observe what is the current state and surface it in conditions, we do not track how do we get to the current state)
- We set reason only when condition have a negative meaning (as far as I remember setting a reason for positive state was considered redoundant, but it was 4 years ago so I might be wrong...)

This PR intentionally does not mean to change what CAPI today is doing WRT to reasons, because this could imply a way bigger impact that probably we should defer to when we cut a new API version.

File: docs/proposals/20200506-conditions.md
Comment (sbueringer) [inline docs/proposals/20200506-conditions.md:247]:
```suggestion
severity is designed to provide a standard classification of possible conditions failure `Reason`. 
```

File: docs/proposals/20200506-conditions.md
Comment (sbueringer) [inline docs/proposals/20200506-conditions.md:476]:
```suggestion
    -  2024-05-03: Change to allow conditions without positive polarity goes into this direction
```
(~ means the same but this framing would also include conditions without polarity. I think Paused is basically neiter positive nor negative)

(same in Implementation History)

File: util/conditions/getter_test.go
Comment (sbueringer) [inline util/conditions/getter_test.go:36]: Is a condition with positive polarity that is false really a negative polarity condition?

(or is the naming here just confusing)

**Review by sbueringer — COMMENTED** [review]
I have to spend more time on it, but NegativeTrueCondition and PositiveFalseCondition are really confusing me

File: util/conditions/setter.go
Comment (theobarberbany) [reply util/conditions/setter.go:111]: > This PR intentionally does not mean to change what CAPI today is doing WRT to reasons, because this could imply a way bigger impact that probably we should defer to when we cut a new API version.

This feels sensible for keeping the scope small here 😄 

For my 0.02c:  this approach 'feels' more right to me (less surprising)

> From a semantic point of view, we use reason to "explain" the current state, not the last transition (in some case the diff is subtle, but in controller we just observe what is the current state and surface it in conditions, we do not track how do we get to the current state)

File: docs/proposals/20200506-conditions.md
Comment (theobarberbany) [reply docs/proposals/20200506-conditions.md:247]: I agree the wording is a touch confusing, I read this as: 

Severity must only be set when something of positive polarity has 'Status=False' (something good is not true), or when something of negative polarity has 'Status =True'  (something bad is true)

**Comment by JoelSpeed** [comment]
What are the next steps here? Do we want to discuss the Positive/Negative nomenclature in more depth? I'm not personally a huge fan of it and would rather users didn't have to worry, ie we could have helpers that allow you just to specify the state true or false explicitly.

But I can also see that this would make a major departure from where we are now and understand the desire to make a small step here to allow more flexibility in conditions. The PR does this in a sympathetic way and unblocks what is required.

I think the naming, and perhaps eventually removing the connotation of positive/negative could be done at a later point once we've understood more about the usage pattern of the negative/neutral condition types being introduced

**Comment by fabriziopandini** [comment]
@JoelSpeed on the positive side there is agreement on the idea to support negative polarity conditions, that's already a good point to be.

The only thing we are discussing are the new helpers: `NegativeTrueCondition`, `PositiveFalseCondition` and the corresponding `MarkNegativeTrue` and `MarkPositiveFalse`

Probably, the most quick way forward is to just drop them, and let users willing to use negative polarity conditions to create Condition types "freely" use the Set(Condition) helper.
I personally think this could be accettable for now because I assume there will be just a few negative polarity conditions at the beginning, and we can come out with better helpers in a second time if necessary.

From the other side of the coin, having helpers really helped users so far in keeping things consistent across all providers, and I still see value in investing some time to figure out new names for helpers. 

Another option, if we all agree that there is value in having helpers, is to start with the current names (or a variation of them we quickly agree upon) and eventually iterate in the future.

**Comment by fabriziopandini** [comment]
Might be using more explicit names for helpers might help:
NegativeTrueCondition --> TrueConditionWithNegativePolarity
PositiveFalseCondition --> FalseConditionWithNegativePolarity
MarkNegativeTrue --> MarkTrueConditionWithNegativePolarity
MarkPositiveFalse --> MarkFalseConditionWithNegativePolarity

(and BTW, the fact that we add helpers don't prevent users to create Condition types "freely" use the Set(Condition) helper)

**Comment by sbueringer** [comment]
Definitely less confusing

File: util/conditions/getter_test.go
Comment (vincepri) [reply util/conditions/getter_test.go:36]: It's very confusing agreed

**Comment by fabriziopandini** [comment]
ACK, I will try get back to this next week.
@enxebre @JoelSpeed are you ok with the proposed way forward?

**Comment by JoelSpeed** [comment]
Ack, I'm happy with the proposed approach forward, please update when you have a chance

**Comment by fabriziopandini** [comment]
@sbueringer @vincepri @JoelSpeed @enxebre @theobarberbany 
renamed the helpers as per comment above/consensus.
PTAL

**Review by JoelSpeed — COMMENTED** [review]
LGTM, will hold off labelling so others can have input

**Comment by theobarberbany** [comment]
Also LGTM :)

**Comment by sbueringer** [comment]
/lgtm
in general

My summary would be
* We now support conditions with positive and negative polarity
* We added util funcs for negative polarity
* For me "Paused" is neither positive nor negative polarity, so I wouldn't use the new util funcs for it, which should be okay
  * Note: Paused should never by part of a summary: `Condition that do not represent the operational state of the component, MUST not be included in the  `Ready` condition  (e.g. `Paused`, which represent a state of the controller that manages the component).`

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
